csaf/src/metadata_retriever_new/mod.rs

```diff
@@ -0,0 +1,259 @@
+use crate::model::metadata::ProviderMetadata;
+use crate::source::HttpSourceError;
+use async_trait::async_trait;
+use sectxtlib::SecurityTxt;
+use url::{ParseError, Url};
+use walker_common::fetcher;
+use walker_common::fetcher::{Fetcher, Json};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Fetch error: {0}")]
+    Fetcher(#[from] fetcher::Error),
+    #[error("URL error: {0}")]
+    Url(#[from] ParseError),
+    #[error("securityTxtParseError error: {0}")]
+    SecurityTextError(#[from] sectxtlib::ParseError),
+    #[error("JSON parse error: {0}")]
+    Json(#[from] serde_json::Error),
+}
+
+impl From<Error> for HttpSourceError {
+    fn from(value: Error) -> Self {
+        match value {
+            Error::Fetcher(err) => Self::Fetcher(err),
+            Error::Url(err) => Self::Url(err),
+            Error::SecurityTextError(err) => Self::SecurityTextError(err),
+            Error::Json(err) => Self::Json(err),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct WellKnownRetriever {
+    pub provider_metadata_url: Url,
+    pub fetcher: Fetcher,
+}
+
+#[derive(Clone)]
+pub struct DnsRetriever {
+    pub provider_metadata_url: Url,
+    pub fetcher: Fetcher,
+}
+
+#[derive(Clone)]
+pub struct SecurityTextRetriever {
+    pub provider_metadata_url: Url,
+    pub fetcher: Fetcher,
+}
+
+#[derive(Clone)]
+pub struct FullSecurityTextRetriever {
+    pub provider_metadata_url: Url,
+    pub fetcher: Fetcher,
+}
+
+#[derive(Clone)]
+pub struct AllApproachRetriever {
+    pub provider_metadata_url: Url,
+    pub fetcher: Fetcher,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum DetectionType {
+    WellKnowPath(ProviderMetadata),
+    DnsPath(ProviderMetadata),
+    SecurityTextPath(ProviderMetadata),
+    FullSecurityTextPath(ProviderMetadata),
+}
+
+#[async_trait(?Send)]
+pub trait Retriever {
+    async fn retrieve(&self) -> Result<DetectionType, Error>
+    where
+        Self: Sized + Send;
+}
+
+async fn get_url_from_security_text(fetcher: &Fetcher, host_url: Url) -> Result<String, Error> {
+    let security_text = fetcher.fetch::<String>(host_url).await?;
+    let text = SecurityTxt::parse(security_text.as_str())?;
+    let url = text
+        .extension
+        .into_iter()
+        .filter(|ext| ext.name == "csaf" && ext.value.contains("https"))
+        .find(|_| true)
+        .map(|ext| ext.value)
+        .unwrap_or_else(|| "".to_string());
```

Comment:
we should not return an empty string. If we don't find a URL, then that should be return as `None`, which means that the return type should be `Option<String>`, or combined with the error: `Result<Option<String>, Error>`.